node/core/bitfield-signing/src/lib.rs

```diff
@@ -49,7 +49,7 @@ use self::metrics::Metrics;
 mod tests;
 
 /// Delay between starting a bitfield signing job and its attempting to create a bitfield.
-const JOB_DELAY: Duration = Duration::from_millis(1500);
+const JOB_DELAY: Duration = Duration::from_millis(1000);
```

Comment:
I think this is a really bad idea. I pointed out this piece of code already 2 years ago. The problem here being that the job starts the moment the block is imported. Which could already be way before the actual slot ends. Meaning we may even start sending out the bitfields before the next slot starts. So, we may also don't yet have imported all chunks.

Reply:
I am not sure I follow. My understanding is the following:

1. We import a block X and find occupied cores.
2. Availability-distribution starts fetching chunks for all those cores
3. After `JOB_DELAY` bitfield signing is checking for available chunks populates bits accordingly and sends them out.
4. Block producer receives bitfields hopefully before [`PRE_PROPOSE_TIMEOUT`](PRE_PROPOSE_TIMEOUT) and includes them in the very next block.
5. Core is free in block X + 1 - candidates have been included, 12 second parachain block time is over.
6. Backing starts and core gets occupied again on block X + 2
7. Repeat at 1.

By reduction of `JOB_DELAY` we increased the time of 0.5s to 1s we have for gossip.

Reply:
Yes, that was actually very enlightening :bulb:  - thanks! So what you are saying is basically that we have more time than we think we have, which is probably true if we were smarter at the provisioner. The provisioner is going to provide the inherent data exactly 2 seconds after it [imported the relay parent](https://github.com/paritytech/polkadot/blob/2c934ed2cddcbf39cc3d8575d086ce13d937ca95/node/core/provisioner/src/lib.rs#L191). If we miss these two seconds, which are counted from block import, votes won't make it. So based on whether the block producer received the relay parent early or late, we would have less or slightly more than 2 seconds time to get the bitfields to the block producer - with `JOB_DELAY` of 1.5 seconds, these are just 0.5 seconds!

Also on Kusama it looks like chunks get usually distributed in less than 0.4 seconds, so reducing the `JOB_DELAY` seems sensible, if bitfields don't make it on time. We seem to be missing bitfields more than bits at the moment.

But what I am really getting here, that we could potentially improve performance significantly if we actually took slots into account in Polkadot. So instead of the provisioner cutting at 2 seconds after block import, it should trigger based on the actual slot.

Reply:
One thing is interesting though. paras_inherent is willing to [wait 2.5 seconds](https://github.com/paritytech/polkadot/blob/4302f98c6acaa376fa176e5557a749d9618ccce2/node/core/parachains-inherent/src/lib.rs#L39) for the provisioner. Assuming that this timeout is correct and we actually have that much budget, this would mean that after a light block, where the block import happened, let's say 2 seconds before the next slot, the provisioner will serve the data immediately, although `paras_inherent` would be willing to wait for 2.5 more seconds! :grimacing:

Reply:
Making the next block "light" again, so the story repeats. And we keep leaving time on the table we could have given backing to actually get candidates in.

Reply:
Yes, we could create more complex block production strategies where we always or at least more often wait until X time within the slot. We could send some final `wait_until: Instant` to the provisioner so it doesn't return immediately. `wait_until` could be based on the time we received the last block and the maximum time we are willing to wait into our slot, and should be structured so we don't _always_ wait maximally into our slot so we don't just shift block production over by 6 seconds.

Reply:
So what we can achieve here is adjusting how stable the time we have for preparing inherent data is and on what condition it should fluctuate in what way. :thinking:

Reply:
We might be able to get better properties, by being more strict and not cutting into the next slot. 
This way, if a parachain does something that is slow (e.g. just big PoVs or issuing a runtime uprade), the following will happen:

1. If the previous relay chain block was rather big, it might not make it.
2. Other smaller stuff is likely ready and can be put into the block.
3. Because of the reduced amount of time we had, the block is likely ending up rather small. (so we have a tendency of producing a small block after a big one)
4. A small block is likely built, distributed and imported fast.
5. We have more time available at the next slot and the big stuff of at least one parachain has good chances of getting in.

This means after a slot with little time, we will have more time in the next slot. This even scales a bit, the less time we have now, the more we will have the next (because the block will be smaller).
So in my mind, by being stricter we might be able to help avoid long stalls and also in general end up with a better distributed load, which should also be beneficial to the system performance.

With the current scheme the following will happen:

A parachains does something slow:

1. If the previous block was rather big, we eat into the slot, still the maximum time budget is 2 seconds in that case, parachain might not make it.
2. Despite the fact that the parachain did not make it (and the block ended up tiny), we just ate into the block, worst case up to 2 seconds - so the time we have for preparing the next block will again be greatly reduced, despite the fact that we did not get much work done with the previous block!

We will need to have a long sequence of small relay chain blocks for this to recover eventually, but the moment the parachain makes it, we will have a big block again and the game starts anew. 

Those are some late night thoughts, but to me that sounds like a promising experiment: Don't eat into the slot.